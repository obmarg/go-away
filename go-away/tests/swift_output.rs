#![allow(dead_code)]

use insta::assert_snapshot;

use go_away::{TypeAlias, TypeMetadata, TypeRegistry};

#[derive(TypeMetadata)]
struct MyData {
    field_one: String,
    nested: Nested,
}

#[test]
fn test_struct_output() {
    let mut registry = TypeRegistry::new();
    MyData::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output::<go_away::SwiftType>(&registry));
}

#[derive(TypeMetadata)]
#[serde(tag = "type", content = "data")]
enum AdjacentlyTaggedTupleEnums {
    OptionOne(One),
    OptionTwo(Two),
}

#[derive(TypeMetadata)]
struct Nested {
    #[serde(rename = "some_other_name")]
    a_string: String,
    an_int: i64,
    fulfilment_type: FulfilmentType,
}

#[derive(TypeMetadata)]
enum FulfilmentType {
    Delivery,
    Collection,
}

#[derive(TypeMetadata)]
struct One {
    x: f32,
}

#[derive(TypeMetadata)]
struct Two {
    y: bool,
}

#[test]
fn test_adjacently_tagged_tuple_enums() {
    let mut registry = TypeRegistry::new();
    AdjacentlyTaggedTupleEnums::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output::<go_away::SwiftType>(&registry));
}

/*
// TODO: Add StructEnum support...

#[derive(TypeMetadata)]
#[serde(tag = "type", content = "data")]
enum StructEnum {
    OptionOne { x: String, y: i32 },
    OptionTwo { foo: String, bar: Nested },
}

#[test]
fn test_struct_enum() {
    let mut registry = TypeRegistry::new();
    StructEnum::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output::<go_away::SwiftType>(&registry));
}
*/

/*

// TODO: Finish this.

#[derive(TypeMetadata)]
#[serde(tag = "type")]
enum InternallyTaggedTupleEnum {
    One(One),
    Two(Two),
}

#[test]
fn test_internally_tagged_tuple_enum() {
    let mut registry = TypeRegistry::new();
    InternallyTaggedTupleEnum::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output::<go_away::SwiftType>(&registry));
}
*/

#[derive(TypeMetadata)]
enum ExternallyTaggedTupleEnum {
    One(One),
    Two(Two),
}

#[test]
fn test_externally_tagged_tuple_enum() {
    let mut registry = TypeRegistry::new();
    ExternallyTaggedTupleEnum::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output::<go_away::SwiftType>(&registry));
}

#[derive(TypeMetadata)]
struct TypeWithLifetimes<'a, 'b> {
    data: &'a str,
    other: &'b str,
}

#[test]
fn lifetimes_and_strs() {
    let mut registry = TypeRegistry::new();
    TypeWithLifetimes::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output::<go_away::SwiftType>(&registry));
}

#[test]
fn type_deduplication() {
    let mut registry = TypeRegistry::new();

    // These both contain `Nested` so there should be one `Nested` type in the output
    //StructEnum::metadata(&mut registry);
    Nested::metadata(&mut registry);
    MyData::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output::<go_away::SwiftType>(&registry));
}

#[test]
fn type_aliases() {
    type MyType = std::collections::HashMap<String, i64>;

    let mut registry = TypeRegistry::new();

    MyType::register_alias("MyType", &mut registry);

    assert_snapshot!(go_away::registry_to_output::<go_away::SwiftType>(&registry), @r###"
    typealias MyType = [String: Int64]
    "###);
}
