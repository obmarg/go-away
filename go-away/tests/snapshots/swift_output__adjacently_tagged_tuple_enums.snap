---
source: go-away/tests/swift_output.rs
expression: "go_away::registry_to_output::<go_away::SwiftType>(&registry)"
---
public struct Two: Hashable, Codable {
    public var y: Bool

    public init(
        y: Bool,
    ) {
        self.y = y
    }
}

extension Two {
    enum CodingKeys: String, CodingKey {
        case y = "y"
    }
}

public struct One: Hashable, Codable {
    public var x: Double

    public init(
        x: Double,
    ) {
        self.x = x
    }
}

extension One {
    enum CodingKeys: String, CodingKey {
        case x = "x"
    }
}

public enum AdjacentlyTaggedTupleEnums {
    case optionOne(One)
    case optionTwo(Two)
    enum CodingKeys: String, CodingKey {
        case optionOne = "OptionOne"
        case optionTwo = "OptionTwo"
    }
}

extension AdjacentlyTaggedTupleEnums: Decodable {
    init(from decoder: Decoder) throws {
        enum ContainerKeys: String, CodingKey {
            case tag = "type"
            case data = "data"
        }
        let container = try decoder.container(ContainerKeys.self)
        let key = try decoder.decode(AdjacentlyTaggedTupleEnums.CodingKeys.self, forKey: .tag)
        switch key {
            case .optionOne:
                return try container.decode(One.self, forKey: .data)
            }
            case .optionTwo:
                return try container.decode(Two.self, forKey: .data)
            }
            default:
                throw "Unknown variant"
        }

    }
}

extension AdjacentlyTaggedTupleEnums: Encodable {
    func encode(to encoder: Encoder) throws {
        enum ContainerKeys: String, CodingKey {
            case tag = "type"
            case data = "data
        }
        let container = try encoder.container(ContainerKeys.self)
        switch this {
            case .optionOne(data):
                try container.encode(AdjacentlyTaggedTupleEnums.CodingKeys.name, forKey: .tag)
                try container.encode(data, forKey: .data)
            }
            case .optionTwo(data):
                try container.encode(AdjacentlyTaggedTupleEnums.CodingKeys.name, forKey: .tag)
                try container.encode(data, forKey: .data)
            }
        }

    }
}



